### PR TITLE
Refactor file storage APIs.

### PR DESCRIPTION
**Pull request by jhendrixMSFT**
Split code for shares, directories and files into their own source files;
the same refactoring was made for the test code as well.
Make CRUD APIs for shares, directories and files methods for their
respective types.
Added support for content MD5 hashes when sending/receiving file ranges.

### CODE REVIEW

**Comment by jhendrixMSFT** [comment]
@mcardosos @marstr Here's the PR for the file storage API refactoring.

**Comment by jhendrixMSFT** [comment]
test this please

**Comment by jhendrixMSFT** [comment]
@azuresdkci test this please

File: storage/file.go [hunk 1]
```diff
@@ -1,118 +1,54 @@
...
-
-// A File is an entry in DirsAndFilesListResponse.
+// File represents a file on a share.
 type File struct {
-	Name       string         `xml:"Name"`
+	fsc        *FileServiceClient
+	Metadata   map[string]string
+	Name       string `xml:"Name"`
+	parent     *Directory
 	Properties FileProperties `xml:"Properties"`
```
Comment (jhendrixMSFT) [inline storage/file.go]: Unfortunately this breaks retrieval of file properties, see the call to headerFromStruct() in SetProperties().  We could probably come up with a solution if we want to make this change however I think we can look at that separately.

File: storage/file.go [hunk 2]
```diff
@@ -133,132 +69,140 @@ type FileRange struct {
...
-	return string(ct)
+	f.updateEtagAndLastModified(headers)
+	return nil
 }
 
-type resourceType string
+// Create creates a new file or replaces an existing one.
+//
+// See https://msdn.microsoft.com/en-us/library/azure/dn194271.aspx
+func (f *File) Create(maxSize uint64) error {
```
Comment (mcardosos) [inline storage/file.go]: Some funcs in the ARM side would name this `CreateOrUpdate`

File: storage/fileserviceclient.go [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+
+	if err = checkRespCode(resp.statusCode, []int{http.StatusOK}); err != nil {
+		resp.body.Close()
+		return nil, err
+	}
+
+	return resp, nil
+}
+
+// returns true if the specified directory or share exists
```
Comment (mcardosos) [inline storage/fileserviceclient.go]: nit... (and file)

File: storage/file.go [hunk 4]
```diff
@@ -393,487 +263,99 @@ func (f FileServiceClient) modifyRange(path string, bytes io.Reader, fileRange F
...
+	if err == nil {
+		f.Properties.Length = size
 	}
-	defer resp.body.Close()
 
-	return checkRespCode(resp.statusCode, []int{http.StatusOK})
+	f.Properties.CacheControl = header.Get("Cache-Control")
+	f.Properties.Disposition = header.Get("Content-Disposition")
+	f.Properties.Encoding = header.Get("Content-Encoding")
+	f.Properties.Etag = header.Get("ETag")
```
Comment (mcardosos) [inline storage/file.go]: Why not add updateEtagAndLastModified instead?

File: storage/file_test.go [hunk 5]
```diff
@@ -374,24 +141,25 @@ func (s *StorageFileSuite) TestFileRanges(c *chk.C) {
...
 
 	// clear partial range and validate
-	c.Assert(cli.ClearRange(file, putRanges[0]), chk.IsNil)
-	c.Assert(cli.ClearRange(file, putRanges[2]), chk.IsNil)
-	ranges, err = cli.ListFileRanges(file, nil)
+	c.Assert(file.ClearRange(putRanges[0]), chk.IsNil)
+	c.Assert(file.ClearRange(putRanges[2]), chk.IsNil)
+	ranges, err = file.ListRanges(nil)
 	c.Assert(err, chk.IsNil)
 	c.Assert(len(ranges.FileRanges), chk.Equals, 2)
```
Comment (mcardosos) [inline storage/file_test.go]: There is also chk.HasLen

File: storage/share.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
+package storage
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"strconv"
+)
+
+// Share represents a Azure file share.
```
Comment (mcardosos) [inline storage/share.go]: nit, "an"

**Comment by mcardosos** [comment]
From my side, just nits :D

File: storage/fileserviceclient.go (hunk 3 above)
Comment (jhendrixMSFT) [inline storage/fileserviceclient.go]: fixed

File: storage/file.go (hunk 4 above)
Comment (jhendrixMSFT) [inline storage/file.go]: fixed

File: storage/file_test.go (hunk 5 above)
Comment (jhendrixMSFT) [inline storage/file_test.go]: fixed

File: storage/share.go (hunk 6 above)
Comment (jhendrixMSFT) [inline storage/share.go]: fixed

**Pull request merged**